cub/cub/grid/grid_even_share.cuh

```diff
@@ -135,8 +135,8 @@ public:
     int avg_tiles_per_block = total_tiles / grid_size;
     // leftover grains go to big blocks:
     this->big_shares         = total_tiles - (avg_tiles_per_block * grid_size);
-    this->normal_share_items = avg_tiles_per_block * tile_items;
-    this->normal_base_offset = big_shares * tile_items;
+    this->normal_share_items = static_cast<OffsetT>(avg_tiles_per_block) * tile_items;
+    this->normal_base_offset = static_cast<OffsetT>(big_shares) * tile_items;
```

Comment:
This would only work if we guarantee that `OffsetT` is an unsigned integer

Reply:
Really wondering whether we should use the safe integer arithmetic that @davebayer is implementing

Reply:
> This would only work if we guarantee that `OffsetT` is an unsigned integer

Could you please elaborate? I see the following types:

```OffsetT = int * int;```

Where `OffsetT` can be `int32`, `uint32`, `int64`, or `uint64`. I could imagine a problem if `OffsetT` can hold more positive numbers than `int`, because the multiplication would truncate before promoting to `OffsetT`.

Reply:
My comment is more that if `OffsetT` is in fact `int`, it does nothing to prevent overflow

Reply:
That's alright. The user chose `OffsetT` and is assigning the product to that type. We cannot prevent that here.